### PR TITLE
adding health checks 

### PR DESCRIPTION
**Pull request by areller**
adding health checks according to this https://github.com/dotnet/tye/issues/19

## Changes
1. Replicas can be cancelled independent of the service
2. There are 2 new configurations for a service: `liveness`, `readiness`. each can hold these settings

```
http:
  path: /health
  port: 80 # default: pick the first binding of the service
  protocol: http # default: none
  headers:
    - name: A
       value: B
initialDelay: 5 # default: 0
period: 1 # default: 1
timeout: 1 # default: 1
successThreshold: 1 # default: 1
failureThreshold: 3 # default: 3
```
3. There are 2 new replica states: `Healthy`, `Ready` (`Ready` = passed both `liveness` and `readiness` probing. `Healthy` = passed `liveness` but not `readiness`).  
If neither `liveness` nor `readiness` were provided, replica will be upgraded to `Ready` upon creation. If only `liveness` if provided, replica will be upgraded to `Ready` upon passing the `liveness` probe. if only `readiness` is provided, replica will be upgraded to `Healthy` upon creation.
4. There is a `ReplicaMonitor` that reads the `liveness` and `readiness` probe configuration, and kills/changes the state of the replica according to probe results
5. Proxy and Ingress only forward to `Ready` replicas
6. The generated Kubernetes manifest contains the `liveness` and `readiness` configuration.

### CODE REVIEW

File: src/Microsoft.Tye.Hosting/DockerRunner.cs [hunk 1]
```diff
@@ -219,210 +219,215 @@ private async Task StartContainerAsync(Application application, Service service,
 
             async Task RunDockerContainer(IEnumerable<(int ExternalPort, int Port, int? ContainerPort, string? Protocol)> ports)
             {
-                var hasPorts = ports.Any();
+                while (!dockerInfo.StoppingTokenSource.IsCancellationRequested)
```
Comment (davidfowl) [inline src/Microsoft.Tye.Hosting/DockerRunner.cs:222]: Why is there a loop here?

Comment (areller) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:222]: When I implement the monitor, it might decide to kill a container because it's unhealthy. In that case I'd want the `DockerRunner` to recreate the replica.
Another way to do it, is to kill the container by calling `docker restart`, and then there is no need for that loop. what do you think?

Comment (areller) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:222]: The argument behind the former, is that it's consistent with how I'll do it in `ProcessRunner`

Comment (davidfowl) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:222]: The docker runner already restarts the replica albiet with the same name.

Comment (areller) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:222]: @davidfowl I think that the problem with doing `docker restart` is that the health monitor won't be able to tell that the container was restarted (because there won't be `Stopped` and `Started` events)

**Comment by areller** [comment]
@davidfowl there are two major things left

1. generating k8s yamls with liveness and readiness
2. I don't forward traffic to non ready replicas, unless there is a single replica. proxying with a single replica causes some tests that use external docker images (e.g. redis) to fail, and I need to check why.

Maybe it's best to start reviewing this PR and I can submit the rest later on this PR or another PR?

File: test/E2ETest/testassets/projects/health-checks/tye-all.yaml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# tye application configuration file
+# read all about it at https://github.com/dotnet/tye
+name: health-checks
+services:
+  - name: all
+    project: api/api.csproj
+    replicas: 3
+    bindings:
+      - port: 8004
+    liveness:
```
Comment (davidfowl) [inline test/E2ETest/testassets/projects/health-checks/tye-all.yaml:10]: These 2 probes are bigger than the entire yaml, that makes me 😢

Comment (areller) [reply test/E2ETest/testassets/projects/health-checks/tye-all.yaml:10]: @davidfowl yeah... I thought it's best to go with something flexible like in k8s, but you can get a way with just providing a `path`, the rest have default values

**Comment by areller** [comment]
I'll probably add the k8s manifest generation tomorrow.

**Comment by jkotalik** [comment]
@areller this change is amazing. Got my coffee ☕  and will be reviewing now.

File: src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs [hunk 3]
```diff
@@ -452,6 +462,64 @@ internal static class KubernetesManifestGenerator
...
+            {
+                var builderHttp = builder.Http!;
+                var httpGet = new YamlMappingNode();
+
+                probe.Add("httpGet", httpGet);
+                httpGet.Add("path", builderHttp.Path);
+
+                if (builderHttp.Protocol != null)
+                {
+                    httpGet.Add("scheme", builderHttp.Protocol!.ToUpper());
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs:480]: I think you don't need `!` here, other similar cases in this file as well.

```suggestion
                    httpGet.Add("scheme", builderHttp.Protocol.ToUpper());
```

File: src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs [hunk 4]
```diff
@@ -452,6 +462,64 @@ internal static class KubernetesManifestGenerator
...
+                }
+
+                if (builderHttp.Port != null)
+                {
+                    httpGet.Add("port", builderHttp.Port!.ToString()!);
+                }
+                else
+                {
+                    // If port is not given, we pull default port
+                    var binding = service.Bindings.First(b => builderHttp.Protocol is null || b.Protocol == builderHttp.Protocol);
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs:490]: It does feel a bit awkward to have health checks per service rather than per endpoint, but it seems to be what k8s does. Or maybe we require a port?

File: src/Microsoft.Tye.Core/ProjectServiceBuilder.cs [hunk 5]
```diff
@@ -54,5 +54,8 @@ public ProjectServiceBuilder(string name, FileInfo projectFile)
         public Dictionary<string, string> BuildProperties { get; } = new Dictionary<string, string>();
 
         public List<SidecarBuilder> Sidecars { get; } = new List<SidecarBuilder>();
+
+        public ProbeBuilder? Liveness { get; set; }
+        public ProbeBuilder? Readiness { get; set; }
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/ProjectServiceBuilder.cs:59]:
```suggestion

        public ProbeBuilder? Readiness { get; set; }
```

File: src/Microsoft.Tye.Hosting/DockerRunner.cs (hunk 1 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:236]: From what I can tell with this loop, you would call `docker rm` and other commands I don't think you want to run: https://github.com/dotnet/tye/blob/65c7ae0d80d0b3749127839732cf7d1ac9926b79/src/Microsoft.Tye.Hosting/DockerRunner.cs#L429

The restart loop for rerunning docker is here: https://github.com/dotnet/tye/blob/65c7ae0d80d0b3749127839732cf7d1ac9926b79/src/Microsoft.Tye.Hosting/DockerRunner.cs#L382. Can you modify this loop instead?

File: src/Microsoft.Tye.Hosting/HttpProxyService.cs [hunk 6]
```diff
@@ -117,23 +122,30 @@ public async Task StartAsync(Application application)
...
 
                         _logger.LogInformation("Service {ServiceName} is using {Urls}", targetServiceDescription.Name, string.Join(",", uris.Select(u => u.ToString())));
 
                         // The only load balancing strategy here is round robin
                         long count = 0;
-                        RequestDelegate del = context =>
+                        RequestDelegate del = async context =>
                         {
                             var next = (int)(Interlocked.Increment(ref count) % uris.Count);
+                            for (int i = 0; !_readyPorts.ContainsKey(uris[next].Port) && i < uris.Count; i++, next = (int)(Interlocked.Increment(ref count) % uris.Count)) ;
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/HttpProxyService.cs:135]: Lot of stuff going on in this for loop, can you break it up a bit and add a comment about what it is doing?

My best guess is loop through all ready ports until we find one that is ready.

File: src/Microsoft.Tye.Hosting/Model/V1/V1ReplicaStatus.cs [hunk 7]
```diff
@@ -18,5 +18,7 @@ public class V1ReplicaStatus
         public int? ExitCode { get; set; }
         public int? Pid { get; set; }
         public IDictionary<string, string>? Environment { get; set; }
+
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/Model/V1/V1ReplicaStatus.cs:21]:
```suggestion
```

File: src/Microsoft.Tye.Hosting/PortAssigner.cs [hunk 8]
```diff
@@ -50,7 +50,7 @@ static int GetNextPort()
                         binding.Port = GetNextPort();
                     }
 
-                    if (service.Description.Replicas == 1)
+                    if ((service.Description.Readiness is null || service.Description.RunInfo is IngressRunInfo) && service.Description.Replicas == 1)
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/PortAssigner.cs:53]: Why add the check for `service.Description.RunInfo is IngressRunInfo`

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+                {
+                    StartLivenessProbe(serviceDesc.Liveness);
+                    StartReadinessProbe(serviceDesc.Readiness);
+                }
+            }
+
+            private void StartLivenessProbe(Probe probe, ReplicaState moveToOnSuccess = ReplicaState.Healthy)
+            {
+                // currently only HTTP is available
+                if (probe.Http is null)
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:129]:
```suggestion
                if (probe.Http == null)
```

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+                {
+                    _logger.LogWarning("cannot start probing replica {name} because probe configuration is not set", _replica.Name);
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:131]:
```suggestion
                    _logger.LogWarning("Cannot start probing replica {name} because probe configuration is not set", _replica.Name);
```

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs (hunk 9 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:129]: And for the rest of the file

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs (hunk 10 above)
Comment (jkotalik) [reply src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:131]: And for the rest of the file.

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+                            break;
+                    }
+                });
+
+                _readinessProber.Start();
+            }
+
+            private void MoveToHealthy(ReplicaState from)
+            {
+                _logger.LogInformation("replica {name} is moving to an healthy state", _replica.Name);
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:233]: These logs should probably be at debug level.

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+                }
+                else
+                {
+                    _logger.LogWarning(message);
+                }
+            }
+
+            public override void Start()
+            {
+                Func<ReplicaBinding, bool> bindingClosure = (_httpProberSettings.Port.HasValue, _httpProberSettings.Protocol != null) switch
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:415]: This probably deserves a comment as well.

File: src/Microsoft.Tye.Hosting/PortAssigner.cs (hunk 8 above)
Comment (areller) [reply src/Microsoft.Tye.Hosting/PortAssigner.cs:53]: because I forgot to remove when added `service.Description.Readiness is null` :)

File: src/Microsoft.Tye.Hosting/HttpProxyService.cs (hunk 6 above)
Comment (areller) [reply src/Microsoft.Tye.Hosting/HttpProxyService.cs:135]: that's right. I'm looping through all the ports (starting from the current selected port, `# of ports` times) to find one that is ready. I'll try to simplify and add comment

File: src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs (hunk 4 above)
Comment (areller) [reply src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs:490]: requiring a port will indeed make it more compatible with k8s, but it will make the `tye.yaml` more verbose and less maintainable in my opinion. one thing that k8s does to help with maintainability is to allow to specify a variable as a port. e.g.

```
port: liveness-port
```

File: src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs (hunk 3 above)
Comment (areller) [reply src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs:480]: yep. I removed it and it didn't throw any warnings. probably because it's inside a block that runs if `builderHttp.Protocol != null`.
It's weird because at times the compiler will throw a warning when not using the `!` operator, even though the expression is in a context where it's guaranteed to not be null...

File: src/Microsoft.Tye.Hosting/DockerRunner.cs (hunk 1 above)
Comment (areller) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs:236]: @jkotalik yeah, makes sense. I'll try doing it that way.

File: src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs (hunk 3 above)
Comment (areller) [reply src/Microsoft.Tye.Core/KubernetesManifestGenerator.cs:480]: found an example: if you go to `ReplicaMonitor.cs`:113 and remove the `!` after `serviceDesc.Readiness`, the compiler will throw a warning, even though there is no way for that variable to be null at that point.

File: src/Microsoft.Tye.Hosting/DockerRunner.cs (hunk 1 above)
Comment (areller) [reply src/Microsoft.Tye.Hosting/DockerRunner.cs]: @jkotalik changed the `DockerRunner` to do `docker restart` when `StoppingTokenSource` is canceled.

File: src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs [hunk 13]
```diff
@@ -187,6 +201,165 @@ private static void HandleServiceVolumes(YamlSequenceNode yamlSequenceNode, List
             }
         }
 
+        private static void HandleServiceProbe(YamlMappingNode yamlMappingNode, ConfigProbe probe)
+        {
+            foreach (var child in yamlMappingNode.Children)
```
Comment (jkotalik) [inline src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs:206]: Can we add some tests for these parsing wise? Much faster for checking these kinds of conditions like making sure initialDelay can't be negative.

File: src/Microsoft.Tye.Hosting/ReplicaMonitor.cs [hunk 14]
```diff
@@ -0,0 +1,455 @@
...
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Tye.Hosting.Model;
+
+namespace Microsoft.Tye.Hosting
+{
+    public class ReplicaMonitor : IApplicationProcessor
+    {
+        private ILogger _logger;
```
Comment (jkotalik) [inline src/Microsoft.Tye.Hosting/ReplicaMonitor.cs:19]: @davidfowl to review this on top of me.

File: src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs [hunk 15]
```diff
@@ -187,6 +201,165 @@ private static void HandleServiceVolumes(YamlSequenceNode yamlSequenceNode, List
             }
         }
 
+        private static void HandleServiceProbe(YamlMappingNode yamlMappingNode, ConfigProbe probe)
```
[the remark this answers is not in the document]
Comment (areller) [reply src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs:204]: Yeah. That would probably make more sense but it will break compatibility with the rest of the methods. What do you think?

File: src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs (hunk 13 above)
Comment (areller) [reply src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs:206]: I'll try to come up with something. You want it as part of this PR? or break it up in a different one?

Comment (jkotalik) [reply src/Microsoft.Tye.Core/Serialization/ConfigServiceParser.cs:206]: I think following the structure here: https://github.com/dotnet/tye/blob/master/test/UnitTests/TyeDeserializationTests.cs#L25.

Mostly just parsing/ validation of tye.yaml to make sure we cover these edge cases correctly.

No preference whether we add them in this PR now or follow up, I'd just like them in general 😄

**Comment by jkotalik** [comment]
I'm going to go ahead and merge this, let's do some follow up on top of it 😄 .

Again, thanks @areller for all the work you have done here

**Comment by areller** [comment]
@jkotalik thank you! Will create a PR in the following days
